beta-operator-rewards/rewards-requirements.ts

```diff
@@ -0,0 +1,529 @@
+    uptimeCoefficient = Math.floor(uptimeCoefficient * HUNDRED);
+
+    requirements.set(IS_UP_TIME_SATISFIED, isUptimeSatisfied);
+
+    /// Pre-params requiremnt
+    // TODO: implement
+
+    /// Version requiremnt
+    // TODO: implement
+
+    // Assemble
+    //   "instances":[
+    //     {
+    //        "uptimePercent":4.251766217084136,
+    //        "preParams":132,
+    //        "version":"2.0.0-1m",
+    //        "ip":"10.102.0.30:9701"
+    //     },
+    //    (...)
+    //    ],
+
+    // console.log("authorizations", authorizations);
+    // console.log("instancesData", instancesData);
+    // console.log("requirements", requirements);
+
+    /// Start assembling peer data and weighted authorizations
+    peerData[stakingProvider] = {
+      applications: Object.fromEntries(authorizations),
+      instances: instancesData,
+      requirements: Object.fromEntries(requirements),
+    };
+
+    // TODO: Add pre-params and version reqs
+    if (
+      requirements.get(IS_BEACON_AUTHORIZED_FACTOR) &&
+      requirements.get(IS_TBTC_AUTHORIZED_FACTOR) &&
+      requirements.get(IS_UP_TIME_SATISFIED)
+    ) {
+      const beacon = BigNumber.from(authorizations.get(BEACON_AUTHORIZATION));
+      const tbct = BigNumber.from(authorizations.get(TBTC_AUTHORIZATION));
+      let minApplicationAuthorization = beacon;
+      if (beacon.gt(tbct)) {
+        minApplicationAuthorization = tbct;
+      }
+      weightedAuthorization[stakingProvider] = {
+        // beaneficiary: <address> TODO: implement
+        weightedAuthorization: minApplicationAuthorization
+          .mul(uptimeCoefficient)
+          .div(HUNDRED)
+          .toString(),
+      };
+      weightedAuthorizations.push(weightedAuthorization);
+    }
+
+    peersData.push(peerData);
+  }
+
+  console.log("peersData: ", JSON.stringify(peersData, null, 2));
+  console.log(
+    "weightedAuthorizations: ",
+    JSON.stringify(weightedAuthorizations, null, 2)
+  );
+}
+
+// Peer uptime requirement. The total uptime for all the instances for a given
+// operator has to be greater than 96% to receive the rewards.
+async function checkUptime(
+  operatorAddress: string,
+  instancesData: any[],
+  rewardsInterval: number
+) {
+  const prometheusAPIQuery = `${prometheusAPI}/query`;
+  // Go back in time relevant to the current date to get data for the exact
+  // rewards interval dates.
+  const offset = Math.floor(Date.now() / 1000) - endRewardsTimestamp;
+  const paramsOperatorUptime = {
+    query: `up{chain_address="${operatorAddress}", job="${prometheusJob}"}
+            [${rewardsInterval}s:${UPTIME_QUERY_RESOLUTION}s] offset ${offset}s`,
+  };
+
+  const instancesByOperator = (
+    await queryPrometheus(prometheusAPIQuery, paramsOperatorUptime)
+  ).data.result;
+
+  // First registered 'up' metric in a given interval <start:end> for a given
+  // operator. Start evaluating uptime from this point.
+  const firstRegisteredUptime = instancesByOperator[0].values[0][0];
```

Comment:
is it always the case that the first instance is also has the first registered uptime?

Reply:
Good point! I checked again on Prometheus, and indeed it looks like the returned instances are not sorted for a given operator. Instances for a given operator can be in a random order (or at least not in ascending order). The first time I looked at the Prometheus graph which had sorted data and I mistakenly made an assumption that returned instances are sorted by their timestamps. However, timestamps within a single instance are sorted from my observation. Added reduce function [f3322a5](https://github.com/keep-network/keep-core/pull/3346/commits/f3322a565c435851cb0a7ec01481034d07b3f7c6)